lib/sshutils/sftp/sftp.go

```diff
@@ -0,0 +1,362 @@
+	srcPaths []string
+	dstPath  string
+	srcFS    FileSystem
+	dstFS    FileSystem
+	opts     Options
+
+	// ProgressWriter is a writer for printing the progress
+	// (used only on the client)
+	ProgressWriter io.Writer
+	// Log optionally specifies the logger
+	Log log.FieldLogger
+}
+
+// FileSystem describes file operations to be done either locally or over SFTP
+type FileSystem interface {
+	// SetContext sets the context for the filesystem used for cancellation
+	SetContext(ctx context.Context)
+	// Type returns whether the filesystem is "local" or "remote"
+	Type() string
+	// Stat returns info about a file
+	Stat(path string) (os.FileInfo, error)
+	// ReadDir returns information about files contained within a directory
+	ReadDir(path string) ([]os.FileInfo, error)
+	// Open opens a file
+	Open(path string) (io.ReadCloser, error)
+	// Create creates a new file
+	Create(path string, mode os.FileMode) (io.WriteCloser, error)
+	// MkDir creates a directory
+	// sftp.Client.Mkdir does not take an os.FileMode, so this can't either
+	Mkdir(path string, mode os.FileMode) error
+	// Chmod sets file permissions
+	Chmod(path string, mode os.FileMode) error
+	// Chtimes sets file access and modification time
+	Chtimes(path string, atime, mtime time.Time) error
+}
+
+// CreateUploadConfig returns a Config ready to upload files
+func CreateUploadConfig(src []string, dst string, opts Options) *Config {
+	c := &Config{
+		srcPaths: src,
+		dstPath:  dst,
+		srcFS:    &localFS{},
+		dstFS:    &remoteFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c
+}
+
+// CreateDownloadConfig returns a Config ready to download files
+func CreateDownloadConfig(src, dst string, opts Options) *Config {
+	c := &Config{
+		srcPaths: []string{src},
+		dstPath:  dst,
+		srcFS:    &remoteFS{},
+		dstFS:    &localFS{},
+		opts:     opts,
+	}
+	c.setDefaults()
+
+	return c
+}
+
+// setDefaults sets default values
+func (c *Config) setDefaults() {
+	logger := c.Log
+	if logger == nil {
+		logger = log.StandardLogger()
+	}
+	c.Log = logger.WithFields(log.Fields{
+		trace.Component: "SFTP",
+		trace.ComponentFields: log.Fields{
+			"SrcPaths":      c.srcPaths,
+			"DstPath":       c.dstPath,
+			"Recursive":     c.opts.Recursive,
+			"PreserveAttrs": c.opts.PreserveAttrs,
+		},
+	})
+}
+
+// TransferFiles transfers files from the configured source paths to the
+// configured destination path over SFTP
+func (c *Config) TransferFiles(ctx context.Context, sshClient *ssh.Client) error {
+	sftpClient, err := sftp.NewClient(sshClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	c.initFS(ctx, sftpClient)
+
+	transferErr := c.transfer(ctx)
+	closeErr := sftpClient.Close()
+	if transferErr != nil {
+		return trace.Wrap(transferErr)
+	}
+	if closeErr != nil {
+		return trace.Wrap(closeErr)
+	}
+	return nil
+}
+
+// initFS ensures the source and destination filesystems are ready to transfer
+func (c *Config) initFS(ctx context.Context, client *sftp.Client) {
```

Comment:
This looks a bit confusing to me. So you pass a client and you set it on every filesystem that is not remote?

Reply:
Just the opposite: give remote filesystems an `sftp.Client`